src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1405,6 +1516,130 @@ private Object writeReplace() {
+            }
+            @SuppressWarnings("unchecked")
+            final K k = (K) key;
+            return stable.orElseSet(new Supplier<V>() {
+                @Override public V get() { return mapper.apply(k); }});
+        }
+
+        @jdk.internal.ValueBased
+        final class StableMapEntrySet extends AbstractImmutableSet<Map.Entry<K, V>> {
+
+            @Stable
+            private final Set<Map.Entry<K, StableValueImpl<V>>> delegateEntrySet;
+
+            StableMapEntrySet() {
+                this.delegateEntrySet = delegate.entrySet();
+            }
+
+            @Override public Iterator<Map.Entry<K, V>> iterator() { return new LazyMapIterator(); }
+            @Override public int                       size() { return delegateEntrySet.size(); }
+            @Override public int                       hashCode() { return StableMap.this.hashCode(); }
+
+            @Override
+            public String toString() {
+                return StableUtil.renderMappings(this, "StableSet", delegateEntrySet, false);
+            }
+
+            @jdk.internal.ValueBased
+            final class LazyMapIterator implements Iterator<Map.Entry<K, V>> {
+
+                @Stable
+                private final Iterator<Map.Entry<K, StableValueImpl<V>>> delegateIterator;
+
+                LazyMapIterator() {
+                    this.delegateIterator = delegateEntrySet.iterator();
+                }
+
+                @Override public boolean hasNext() { return delegateIterator.hasNext(); }
+
+                @Override
+                public Entry<K, V> next() {
+                    final Map.Entry<K, StableValueImpl<V>> inner = delegateIterator.next();
+                    final K k = inner.getKey();
+                    return new NullableKeyValueHolder<>(k, inner.getValue().orElseSet(new Supplier<V>() {
+                        @Override public V get() { return mapper.apply(k); }}));
+                }
+
+                @Override
+                public void forEachRemaining(Consumer<? super Map.Entry<K, V>> action) {
+                    final Consumer<? super Map.Entry<K, StableValueImpl<V>>> innerAction =
+                            new Consumer<>() {
+                                @Override
+                                public void accept(Entry<K, StableValueImpl<V>> inner) {
+                                    final K k = inner.getKey();
+                                    action.accept(new NullableKeyValueHolder<>(k, inner.getValue().orElseSet(new Supplier<V>() {
+                                        @Override public V get() { return mapper.apply(k); }})));
+                                }
+                            };
+                    delegateIterator.forEachRemaining(innerAction);
+                }
+            }
+        }
+
+        @Override
+        public Collection<V> values() {
+            return new StableMapValues();
+        }
+
+        final class StableMapValues extends AbstractCollection<V> {
```

Comment:
Can this be `AbstractImmutableCollection`? I know we have #18522 debate going on, but we can probably safely port the semantics introduced in that PR to new collections we add in this patch.